lib/Dialect/FIRRTL/Transforms/GrandCentralSignalMappings.cpp

```diff
@@ -0,0 +1,322 @@
+/// `SignalDriverAnnotation` to the list of mappings that need to be established
+/// for this module.
+void ModuleSignalMappings::addTargets(ArrayAttr targets, MappingDirection dir) {
+  const char *dirString =
+      dir == MappingDirection::DriveRemote ? "sink" : "source";
+  for (auto target : targets.getAsRange<DictionaryAttr>()) {
+    SignalMapping mapping;
+    mapping.dir = dir;
+    mapping.remoteTarget = target.getAs<StringAttr>("_1");
+    if (!mapping.remoteTarget) {
+      module.emitError("SignalDriverAnnotation ")
+          << dirString << " target " << target << " missing \"_1\" attribute";
+      anyFailed = true;
+      continue;
+    }
+    auto localTarget = target.getAs<StringAttr>("_2");
+    if (!localTarget) {
+      module.emitError("SignalDriverAnnotation ")
+          << dirString << " target " << target << " missing \"_2\" attribute";
+      anyFailed = true;
+      continue;
+    }
+    mapping.localName = localTarget.getValue();
+    LLVM_DEBUG(llvm::dbgs() << "  - " << mapping << "\n");
+    mappings.push_back(std::move(mapping));
+  }
+}
+
+void ModuleSignalMappings::emitMappingsModule() {
+  LLVM_DEBUG(llvm::dbgs() << "- Generating `" << mappingsModuleName << "`\n");
+  auto *context = module.getContext();
+
+  // Determine what ports this module will have, given the signal mappings we
+  // are supposed to emit.
+  SmallVector<ModulePortInfo> ports;
+  for (auto &mapping : mappings) {
+    // TODO: We really need the local operation this is targeting, to get at the
+    // type and all the other good stuff.
+    ports.push_back(ModulePortInfo{
+        StringAttr::get(context, mapping.localName),
+        // TODO: This needs to come from the local target:
+        mapping.type,
+        mapping.dir == MappingDirection::DriveRemote ? Direction::Input
+                                                     : Direction::Output,
+        module.getLoc()});
+    LLVM_DEBUG(llvm::dbgs()
+               << "  - Adding port `" << mapping.localName << "`\n");
+  }
+
+  // Create the actual module.
+  ImplicitLocOpBuilder builder(module.getLoc(), module);
+  auto mappingsModule = builder.create<FModuleOp>(
+      StringAttr::get(module.getContext(), mappingsModuleName), ports);
+
+  // Generate the connect and force statements inside the module.
+  builder.setInsertionPointToStart(mappingsModule.getBodyBlock());
+  unsigned portIdx = 0;
+  for (auto &mapping : mappings) {
+    // TODO: Actually generate a proper XMR here. For now just do some textual
+    // replacements.
+    auto circuitSplit = mapping.remoteTarget.getValue().split('|').second;
+    auto moduleSplit = circuitSplit.split('>');
+    SmallString<32> remoteXmrName(moduleSplit.first.split(':').first);
+    remoteXmrName.push_back('.');
+    for (auto c : moduleSplit.second) {
+      if (c == '[' || c == '.')
+        remoteXmrName.push_back('_');
+      else if (c != ']')
+        remoteXmrName.push_back(c);
+    }
```

Comment:
This is doing it this way and not with scattering because the scattering is across multiple circuits, correct?

Reply:
Once we have converged on XMRs being yet another case of Non-Local Annotations (and have some prototype on a simpler pass, like GCT data taps), it'll be easy to go back in and clean this pass up. Going to mark this as resolved because that's probably going to take some more time and the current impl is on the same level of compatibility as GCT data taps.